graphql_compiler/schema_merging/rename_schema.py

```diff
@@ -0,0 +1,232 @@
+                         reverse_root_field_map=reverse_root_field_map)
+
+
+def _rename_types(ast, rename_func, query_type, scalars):
+    """Rename types, enums, interfaces using rename_func.
+
+    The query type will not be renamed. Scalar types, field names, enum values will not be renamed.
+
+    ast will be modified as a result.
+
+    Args:
+        ast: Document, the schema ast that we modify
+        rename_func: callable, used to rename types, interfaces, enums, etc
+        query_type: string, name of the query type, e.g. 'RootSchemaQuery'
+        scalars: set of strings, the set of user defined scalars
+
+    Returns:
+        Dict[str, str], the new type name to original type name map
+
+    Raises:
+        SchemaError if the rename causes name conflicts
+    """
+    visitor = RenameSchemaVisitor(rename_func, query_type, scalars)
+    visit(ast, visitor)
+
+    return visitor.reverse_name_map
+
+
+def _rename_root_fields(ast, rename_func, query_type):
+    """Rename root fields, aka fields of the query type.
+
+    ast will be modified as a result.
+
+    Args:
+        ast: Document, the schema ast that we modify
+        rename_func: callable, used to rename fields of the query type
+        query_type: string, name of the query type, e.g. 'RootSchemaQuery'
+
+    Returns:
+        Dict[str, str], the new root field name to original root field name map
+
+    Raises:
+        SchemaError if rename causes root field name conflicts
+    """
+    visitor = RenameRootFieldsVisitor(rename_func, query_type)
+    visit(ast, visitor)
+
+    return visitor.reverse_field_map
+
+
+class RenameSchemaVisitor(Visitor):
+    """Traverse a Document AST, editing the names of nodes."""
+    def __init__(self, rename_func, query_type, scalar_types):
+        self.rename_func = rename_func  # callable that takes string to string
+        self.reverse_name_map = {}  # Dict[str, str], from new name to original name
+        self.query_type = query_type  # str
+        self.scalar_types = frozenset(scalar_types)
+        self.builtin_types = frozenset({'String', 'Int', 'Float', 'Boolean', 'ID'})
+
+        # NOTE: behavior of Directive type is not completely clear
+        self.noop_types = frozenset({
+            'Name', 'Document', 'Argument', 'IntValue', 'FloatValue', 'StringValue',
+            'BooleanValue', 'EnumValue', 'ListValue', 'Directive', 'ListType', 'NonNullType',
+            'SchemaDefinition', 'OperationTypeDefinition', 'ScalarTypeDefinition',
+            'FieldDefinition', 'InputValueDefinition', 'EnumValueDefinition',
+            'DirectiveDefinition'
+        })
+        self.rename_types = frozenset({
+            'NamedType', 'ObjectTypeDefinition', 'InterfaceTypeDefinition',
+            'UnionTypeDefinition', 'EnumTypeDefinition'
+        })
+        self.unexpected_types = frozenset({
+            'OperationDefinition', 'SelectionSet', 'Field', 'FragmentSpread',
+            'InlineFragment', 'FragmentDefinition'
```

Comment:
Nit: Imagine needing to add one more entry to one of these frozensets. The produced diff will show "removed one row, added one row" and will likely not make it obvious which entry was added.

As an alternative, consider the following format:
```
self.unexpected_types = frozenset({
    'Field',
    'FragmentDefinition',
    'FragmentSpread',
    'InlineFragment',
    'OperationDefinition',
    'SelectionSet',
})
```

This gets us the nice property that any one addition or removal is a single-line operation that affects only the entry being added or removed. Note the trailing comma after the last entry, which makes sure this is the case even for new entries that are at the end of the sequence.